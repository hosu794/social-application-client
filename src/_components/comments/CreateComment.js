import React from "react";
import { Formik, Form, Field, ErrorMessage } from "formik";
import * as Yup from "yup";
import { useSelector, useDispatch } from "react-redux";
import { commentActions } from "../../_actions";
import PropTypes from "prop-types";

function CreateComment({ id }) {
  const dispatch = useDispatch();

  return (
    <section className="column">
      <div className="card">
        <Formik
          className="hero-body"
          initialValues={{
            body: "",
          }}
          validationSchema={Yup.object().shape({
            body: Yup.string().required("Text is required"),
          })}
          onSubmit={(fields) => {
            dispatch(
              commentActions.createComment({
                body: fields.body,
                storyId: id,
              })
            );
          }}
          render={({ errors, status, touched }) => (
            <Form>
              <h1
                className="title"
                style={{
                  color: "hsl(0, 0%, 21%)",
                }}
              >
                Write a comment
              </h1>

              <div className="form-group">
                <Field
                  placeholder="Write some text..."
                  name="body"
                  type="body"
                  className={
                    "input" + (errors.body && touched.body ? " is-danger" : "")
                  }
                />
                <ErrorMessage
                  name="body"
                  component="div"
                  className="help is-danger"
                />
              </div>

              <div className="form-group">
                <button
                  type="submit"
                  className={"button is-info is-medium "}
                  style={{
                    margin: "2rem 0 2rem 0",
                  }}
                >
                  Submit
                </button>
              </div>
            </Form>
          )}
        />
      </div>
    </section>
  );
}

CreateComment.propTypes = {
  id: PropTypes.number.isRequired,
};

export default CreateComment;
