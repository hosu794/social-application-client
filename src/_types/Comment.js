import PropTypes from "prop-types";
import { Component } from "react";
import { IUser } from "./";

export const IComment = (Component.PropTypes = {
  id: PropTypes.number.isRequired,
  body: PropTypes.string.isRequired,
  createdBy: PropTypes.instanceOf(IUser),
});
