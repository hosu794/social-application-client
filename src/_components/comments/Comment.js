import React from "react";

import { useDispatch, useSelector } from "react-redux";
import DeleteCommentButton from "./DeleteCommentButton";
import EditCommentButton from "./EditCommentButton";
import { commentActions } from "../../_actions";
import PropTypes from "prop-types";
import { IComment, IUser } from "../../_types";

function Comment({ comment }) {
  const dispatch = useDispatch();

  const user = useSelector((state) => state.user.user);
  const isUserExist = user ? user.id : null;
  const isUserIdentificationEqualStoryCreator =
    isUserExist === comment.createdBy.id;

  function handleDeleteComment() {
    dispatch(commentActions.deleteComment(comment.id));
  }

  return (
    <div className="column">
      <div className="card">
        <nav className="level">
          <div class="level-item has-text-centered">
            <div className="is-size-5">{comment.body}</div>
          </div>
        </nav>
        <nav className="level">
          <div class="level-item has-text-centered">
            <div className="is-size-6">
              created by <b>{comment.createdBy.username}</b>
            </div>
          </div>
          <div className="level-item has-text-centered">
            {isUserIdentificationEqualStoryCreator ? (
              <DeleteCommentButton handleDeleteComment={handleDeleteComment} />
            ) : null}
          </div>
          <div className="level-item has-text-centered">
            {isUserIdentificationEqualStoryCreator ? (
              <EditCommentButton id={comment.id} />
            ) : null}
          </div>
        </nav>
      </div>
    </div>
  );
}

Comment.propTypes = {
  comment: PropTypes.instanceOf(IComment).isRequired,
  user: PropTypes.instanceOf(IUser),
  isUserExist: PropTypes.number,
  isUserIdentificationEqualStoryCreator: PropTypes.bool,
};

Comment.defaultProps = {
  comment: null,
  user: {},
  isUserExist: undefined,
  isUserIdentificationEqualStoryCreator: false,
};

export default Comment;
