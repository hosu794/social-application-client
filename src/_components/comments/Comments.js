import React, { useEffect, useState } from "react";

import { useDispatch, useSelector } from "react-redux";
import { commentActions } from "../../_actions";
import Comment from "./Comment";
import CreateComment from "./CreateComment";
import PropTypes from "prop-types";
import { IComment } from "../../_types";

function Comments({ id }) {
  const dispatch = useDispatch();

  const [currentPage, setCurrentPage] = useState(0);
  const content = useSelector((state) => state.comment.content);
  const loading = useSelector((state) => state.comment.loading);
  const loggedIn = useSelector((state) => state.authentication.loggedIn);
  const isStoryHasComments = Array.isArray(content) && content.length;

  const getCommentsByStoryId = async () => {
    await dispatch(commentActions.getCommentsByStoryId(id, currentPage));
  };

  useEffect(() => {
    getCommentsByStoryId();
  }, [currentPage]);

  return (
    <section class="hero">
      <div class="hero-body">
        <div class="container">
          {loggedIn ? (
            <div>
              <CreateComment id={id} />
            </div>
          ) : null}
          <h1 class="title">Comments</h1>
          {!loading ? content.map((item) => <Comment comment={item} />) : null}
          {!isStoryHasComments ? (
            <p className="is-size-4">Story hasn't comments!</p>
          ) : null}
        </div>
      </div>
    </section>
  );
}

Comments.propTypes = {
  id: PropTypes.number.isRequired,
  content: PropTypes.arrayOf(IComment),
  loading: PropTypes.bool,
  loggedIn: PropTypes.bool,
  isStoryHasComments: PropTypes.bool,
};

Comment.defaultProps = {
  id: null,
  content: [],
  loading: false,
  loggedId: false,
  isStoryHasComments: false,
};

export default Comments;
