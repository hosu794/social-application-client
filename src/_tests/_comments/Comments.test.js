import React from "react";

import configureStore from "redux-mock-store";
import { useDispatch } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import Comment from "../../_components/comments/Comment";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import Comments from "../../_components/comments/Comments";
import CreateComment from "../../_components/comments/CreateComment";

const mockStore = configureStore([]);

let store = mockStore({
  comment: {
    content: [
      {
        id: 1,
        body: "Some body",
        createdBy: {
          id: 12,
          name: "Joe Doe",
          username: "exampleusername",
          downloadAvatar: "dsdasdasdasdasdadas.png",
        },
      },
    ],
  },
  user: {
    isUverLovedStory: true,
    user: {
      id: 12,
      name: "Joe Doe",
      username: "exampleusername",
      downloadAvatar: "dsdasdasdasdasdadas.png",
    },
  },
  authentication: {
    loggedIn: true,
  },
  stories: {
    content: [],
    currentStory: {
      id: 1,
      title: "Story Title",
      description: "Story description",
      totalLoves: 12,
      body: "<p>Simple Story</p>",
      createdBy: {
        id: 12,
        name: "Joe Doe",
        username: "exampleusername",
        downloadAvatar: "dsdasdasdasdasdadas.png",
      },
      topic: {
        id: 12,
        title: "Topic Example Title",
        description: "Description Example description",
        createdAt: "Dummy date",
        createdBy: {
          id: 12,
          name: "Joe Doe",
          username: "exampleusername",
          downloadAvatar: "dsdasdasdasdasdadas.png",
        },
      },
    },
  },
});

let mockComment = {
  id: 1,
  body: "Some body",
  createdBy: {
    id: 12,
    name: "Joe Doe",
    username: "exampleusername",
    downloadAvatar: "dsdasdasdasdasdadas.png",
  },
};

describe("Test for the Comments Components", () => {
  let TestingComponent = () => (
    <MemoryRouter>
      <Provider store={store}>
        <Comments id={12} />
      </Provider>
    </MemoryRouter>
  );

  let wrapper = mount(<TestingComponent />);

  test("should render component correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should component be defined", () => {
    expect(Comments).toBeDefined();
  });

  test("should doesn't render Paragraph with text", () => {
    let element = wrapper.exists(".is-size-4");

    expect(element).toEqual(false);
  });

  test("should render CreateComment component", () => {
    let element = wrapper.exists(CreateComment);

    expect(element).toEqual(true);
  });

  test("should render list of comments", () => {
    let element = wrapper.exists(Comment);

    expect(element).toEqual(true);
  });
});
