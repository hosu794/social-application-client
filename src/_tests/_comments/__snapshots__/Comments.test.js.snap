// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the Comments Components should render component correctly 1`] = `ReactWrapper {}`;
